apps/shinkai-desktop/src/components/playground-tool/components/code-panel.tsx

```diff
@@ -0,0 +1,181 @@
+import { ToolMetadata } from '@shinkai_network/shinkai-message-ts/api/tools/types';
+import { Button, Skeleton } from '@shinkai_network/shinkai-ui';
+import { debounce } from '@shinkai_network/shinkai-ui/helpers';
+import { cn } from '@shinkai_network/shinkai-ui/utils';
+import { AnimatePresence, motion } from 'framer-motion';
+import { memo, MutableRefObject } from 'react';
+import { useFormContext } from 'react-hook-form';
+
+import { usePlaygroundStore } from '../context/playground-context';
+import { useAutoSaveTool } from '../hooks/use-create-tool-and-save';
+import { CreateToolCodeFormSchema } from '../hooks/use-tool-code';
+import ToolCodeEditor from '../tool-code-editor';
+import { detectLanguage } from '../utils/code';
+
+export const getRandomWidth = () => {
+  const widths = [
+    'w-12',
+    'w-16',
+    'w-20',
+    'w-24',
+    'w-32',
+    'w-40',
+    'w-48',
+    'w-56',
+    'w-64',
+    'w-72',
+  ];
+  return widths[Math.floor(Math.random() * widths.length)];
+};
+
+function CodePanelBase({
+  handleApplyChangesCodeSubmit,
+  resetToolCode,
+  isDirtyCodeEditor,
+  setIsDirtyCodeEditor,
+  baseToolCodeRef,
+}: {
+  handleApplyChangesCodeSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
+  resetToolCode: () => void;
+  isDirtyCodeEditor: boolean;
+  setIsDirtyCodeEditor: (isDirty: boolean) => void;
+  baseToolCodeRef: MutableRefObject<string>;
+}) {
+  const resetCounter = usePlaygroundStore((state) => state.resetCounter);
+  const toolMetadata = usePlaygroundStore((state) => state.toolMetadata);
+  const toolCode = usePlaygroundStore((state) => state.toolCode);
+  const toolCodeStatus = usePlaygroundStore((state) => state.toolCodeStatus);
+  const form = useFormContext<CreateToolCodeFormSchema>();
+  const codeEditorRef = usePlaygroundStore((state) => state.codeEditorRef);
+  const { handleAutoSave } = useAutoSaveTool();
+
+  const isToolCodeGenerationPending = toolCodeStatus === 'pending';
+  const isToolCodeGenerationSuccess = toolCodeStatus === 'success';
+
+  const handleCodeUpdate = debounce((currentCode: string) => {
```

Comment:
Perhaps it's better to use useCallback here to prevent unnecessary re-creations of the debounced function